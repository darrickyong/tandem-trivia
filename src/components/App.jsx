import React, { useState } from "react";
import { FaLinkedin, FaGithubAlt, FaAngellist } from "react-icons/fa";
import Game from "./Game";
import Error from "./Error";
import defaultData from "../Apprentice_TandemFor400_Data.json";

function App() {
  const [apiLink, setApiLink] = useState("");
  const [triviaData, setTriviaData] = useState([]);
  const [errors, setErrors] = useState([]);
  const [highScore, setHighScore] = useState(localStorage.highScore || 0);

  const checkJSON = sampleQuestion => {
    let question = typeof sampleQuestion.question == "string";
    let correct = typeof sampleQuestion.correct == "string";
    let incorrect = sampleQuestion.incorrect.every( choice => typeof choice == "string") && sampleQuestion.incorrect.length === 3;
    return question && correct && incorrect;
  }  

  const handleUpload = e => {
    let reader = new FileReader();
    reader.onload = onReaderLoad;
    if (e.target.files[0].type !== "application/json") {
      const fileType = e.target.files[0].type.split("/")[1].toUpperCase();
      setErrors([`You uploaded a ${fileType} file. Please upload a JSON file instead.`])
      return;
    }
    reader.readAsText(e.target.files[0]);
  }
  
  const onReaderLoad = e => {
    const triviaData = JSON.parse(e.target.result);
    if (checkJSON(triviaData[0])) {
      setErrors([]);
      setTriviaData(triviaData);
    } else {
      setErrors(["Please check that the JSON format is correct."])
    }
  }

  const fetchData = link => {
    if(link.length) {
      fetch(link)
        .then(res => res.json())
        .then(data => {
          if (checkJSON(data[0])) {
            setErrors([]);
            setTriviaData(data);
          } else {
            setErrors(["Please check that the JSON format is correct."])
          }
        })
    }
  }

  const quickStart = e => {
    setTriviaData(defaultData);
  }

  const clearHighScore = () => {
    setHighScore(0);
    localStorage.highScore = 0;
  }

  const welcome = (
    <div className="welcome">
      <div className="welcome-high">
        {`High Score: ${highScore}`}
        <div
          className="welcome-high-button"
          onClick={clearHighScore}>Clear High Score</div>
      </div>
      <h1>Tandem Trivia Game!</h1>
      <div className="welcome-buttons">
        <div
          className="welcome-button" 
          onClick={quickStart}
        >Quickstart</div>

        <label 
          className="welcome-button"
          htmlFor="file"
        > Upload JSON File 
          <input 
            className="upload"
            id="file"
            type="file" 
            name="file"
            onChange={handleUpload}
          />
        </label>

        <div 
          className="welcome-button"
          onClick={() => fetchData(apiLink)}
        >
          <input
            className="api" 
            type="text"
            placeholder="Paste JSON API link here"
            value={apiLink}
            onChange={e => setApiLink(e.target.value)}/>
          Fetch from API
        </div>
      </div>
    </div>
  )

  const socialMedia = (
    <div className="social">
      <a href="https://www.linkedin.com/in/darrick-yong/">
        <FaLinkedin className="linkedin" />
      </a>

      <a href="https://github.com/darrickyong">
        <FaGithubAlt className="github" />
      </a>

      <a href="https://angel.co/u/darrick-yong">
        <FaAngellist className="angel" />
      </a>
    </div>
  )

  return (
    <div className="app">
      {socialMedia}
      {triviaData.length ?  
        <Game 
          triviaData={triviaData} 
          setTriviaData={setTriviaData}
          highScore={highScore}
          setHighScore={setHighScore}
        /> :
        welcome}
      <Error errors={errors} />
    </div>
  );
}

export default App;