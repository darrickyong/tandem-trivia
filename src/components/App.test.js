import React from "react";
import ReactDOM from 'react-dom'
import ShallowRenderer from 'react-test-renderer/shallow';
import renderer from 'react-test-renderer';
import { shallow } from "enzyme";
import App from "./App";

describe('My App', () => {
  
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
  });
  
  it('renders correctly', () => {
    const app = renderer.create(<App />).toJSON();
    expect(app).toMatchSnapshot();
  });

  it('should render correctly with no props', () => {
    const component = shallow(<App />);
    expect(component).toMatchSnapshot();
  });
});