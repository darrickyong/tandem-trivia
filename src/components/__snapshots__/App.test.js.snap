// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`My App should render correctly with no props 1`] = `
<div
  className="app"
>
  <div
    className="social"
  >
    <a
      href="https://www.linkedin.com/in/darrick-yong/"
    >
      <FaLinkedin
        className="linkedin"
      />
    </a>
    <a
      href="https://github.com/darrickyong"
    >
      <FaGithubAlt
        className="github"
      />
    </a>
    <a
      href="https://angel.co/u/darrick-yong"
    >
      <FaAngellist
        className="angel"
      />
    </a>
  </div>
  <div
    className="welcome"
  >
    <div
      className="welcome-high"
    >
      High Score: 0
      <div
        className="welcome-high-button"
        onClick={[Function]}
      >
        Clear High Score
      </div>
    </div>
    <h1>
      Tandem Trivia Game!
    </h1>
    <div
      className="welcome-buttons"
    >
      <div
        className="welcome-button"
        onClick={[Function]}
      >
        Quickstart
      </div>
      <label
        className="welcome-button"
        htmlFor="file"
      >
         Upload JSON File
        <input
          className="upload"
          id="file"
          name="file"
          onChange={[Function]}
          type="file"
        />
      </label>
      <div
        className="welcome-button"
        onClick={[Function]}
      >
        <input
          className="api"
          onChange={[Function]}
          placeholder="Paste JSON API link here"
          type="text"
          value=""
        />
        Fetch from API
      </div>
    </div>
  </div>
  <Error
    errors={Array []}
  />
</div>
`;
